llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h

```diff
@@ -192,6 +191,47 @@ using AllBinaryRecipe_match =
     BinaryRecipe_match<Op0_t, Op1_t, Opcode, VPWidenRecipe, VPReplicateRecipe,
                        VPWidenCastRecipe, VPInstruction>;
 
+template <typename Op0_t, typename Op1_t, unsigned Opcode, bool Commutable,
+          typename... RecipeTys>
+struct LogicalRecipe_match {
+  Op0_t LHS;
+  Op1_t RHS;
+
+  LogicalRecipe_match(Op0_t LHS, Op1_t RHS) : LHS(LHS), RHS(RHS) {}
+
+  bool match(const VPValue *V) {
+    auto *DefR = V->getDefiningRecipe();
+    return DefR && match(DefR);
+  }
+
+  bool match(const VPSingleDefRecipe *R) {
+    return match(static_cast<const VPRecipeBase *>(R));
+  }
+
+  bool match(const VPRecipeBase *R) {
+    if (!detail::MatchRecipeAndOpcode<Opcode, RecipeTys...>::match(R)) {
+      if (!detail::MatchRecipeAndOpcode<Instruction::Select,
+                                        RecipeTys...>::match(R))
+        return false;
+      if (Opcode == Instruction::And) {
+        if (!m_False().match(R->getOperand(2)))
```

Comment:
This use of m_False applies to a general bitwise AND, rather than logical AND of boolean mask operands. But the redundancy holds if the operand is all-false, rather than restricting to a single zero bit.

Reply:
Not sure if I am missing something, but this here doesn't match the opcode of the recipe (that is matched above to be a select), it matches the operation that the matcher was instantiated to match.

Reply:
This stems from searching for `select i1 %c, i1 %a, i1 0` as the poison-preserving form of `and i1 %c, %a`? Would it be clearer for VPlan to represent Poison-Preserving AND/OR directly, possibly via a flag, whose execute() emits the desired select patterns, rather than reverse-engineer them?

Reply:
That would be an option (i.e. add `LogicalAnd/LogicalOr` VPInstruciton opcodes or as separate recipes), but would possibly require performing this canonicalization fora all inputs? 

In any case, better as separate patch, if desired?

Reply:
Would doing so earlier simplify this "Simplify (X && Y) || (X && !Y) -> X" patch? Regarding handling input, is that needed or expected, given that this tries to simplify mask computations that are generated earlier by LV. This is also consistent with using and matching Not (rather than its xor-with-true form), introduced before VPlan represented constants.

Reply:
Put up to add a logical-and opcode and use that when creating edge masks: https://github.com/llvm/llvm-project/pull/91897

> Independently, should createBlockInMask() generate logical OR's using selects as well, possibly by recipes that target IRBuilder::CreateLogicalOr(), rather than plain bitwise OR's via IRBuilder's CreateBinOp()?

Hmm, I am not sure. Would probably warrant a separate look.

> The only AND/OR VPInstructions currently in use are those generated by create[Edge,BlockIn]Mask(), which are logical(ly dealing with masks), right? Actually, VPBuilder::createAnd() appears to be useless, after createEdgeMask() switched to calling VPBuilder::createSelect() instead, so no AND VPInstructions appear to exist, neither bitwise nor logical.

I think so yes, let me clean this up separately.

Reply:
Updated now to be based on LogicalAnd. 

> Perhaps Not should also become a LogicalNot, generating a select %c, false, true instead of the natural xor %c, true? 

I am not entirely sure about changing  the IR we generate here, the XOR variant appears to be the canonical for expected by other passes (i.e. that's what IRBuilder.CreateNot generates)

Reply:
>> Independently, should createBlockInMask() generate logical OR's using selects as well, possibly by recipes that target IRBuilder::CreateLogicalOr(), rather than plain bitwise OR's via IRBuilder's CreateBinOp()?

> Hmm, I am not sure. Would probably warrant a separate look.

Looking at the original issue https://bugs.llvm.org/show_bug.cgi?id=48832 and its fix, it appears that masks of unreachable edges can no longer convey poison, so ORing them to produce block-in-masks should continue to be bitwise rather than logical. OTOH, logical ORs require their operands to be suitably ordered, whereas predecessors are in general unordered (unlike employing a logical AND for edge masks where the mask of the block naturally precedes the condition of its terminal branch).


>> Perhaps Not should also become a LogicalNot, generating a select %c, false, true instead of the natural xor %c, true?

> I am not entirely sure about changing the IR we generate here, the XOR variant appears to be the canonical for expected by other passes (i.e. that's what IRBuilder.CreateNot generates)

The condition of a terminal branch may be poison, so may its negating NOT, provided the edge mask uses a logical AND to avoid propagating it in case the block is unreachable (has a false block-in-mask).


Note: better eliminate dead edges and nodes from the CFG prior to applying predication, once this is done via VPlan-to-VPlan transformation. Doing so may allow to use bitwise ANDs rather than logical ones, as before.